### PR TITLE
enh(API): Refacto PlatformInformation

### PR DESCRIPTION
**Pull request by jeremyjaouen**
## Description

This PR is a refacto of the PlatformInformation entity and the PATCH platform endpoint, using the new architecture

**Fixes** # MON-6552

## Type of change

- [x] Patch fixing an issue (non-breaking change)
- [ ] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software
- [ ] Updating documentation (missing information, typo...)

## Target serie

- [ ] 19.10.x
- [ ] 20.04.x
- [x] 20.10.x
- [x] 21.04.x (master)

## Checklist

- [x] I have followed the **coding style guidelines** provided by Centreon
- [x] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [x] I have commented my code, especially **hard-to-understand areas** of the PR.
- [ ] I have made corresponding changes to the **documentation**.
- [x] I have **rebased** my development branch on the base branch (master, maintenance).

### CODE REVIEW

File: src/Centreon/Domain/PlatformInformation/Interfaces/DtoValidatorInterface.php [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+ */
+declare(strict_types=1);
+
+namespace Centreon\Domain\PlatformInformation\Interfaces;
+
+interface DtoValidatorInterface
+{
+    /**
+     * @param array<string,mixed> $dto
+     * @return void
```
Comment (jeremyjaouen) [inline src/Centreon/Domain/PlatformInformation/Interfaces/DtoValidatorInterface.php:29]: add @throws Throwable

File: src/Centreon/Domain/PlatformInformation/UseCase/V20/UpdatePartiallyPlatformInformation.php [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+    private function addValidator(DtoValidatorInterface $dtoValidator): void
+    {
+        $this->validators[] = $dtoValidator;
+    }
+
+    /**
+     * Execute the use case for which this class was designed.
+     *
+     * @param array<string,mixed> $request
+     * @throws \Exception
```
Comment (jeremyjaouen) [inline src/Centreon/Domain/PlatformInformation/UseCase/V20/UpdatePartiallyPlatformInformation.php:111]: @throws Throwable

File: src/Centreon/Domain/PlatformInformation/Model/PlatformInformationFactory.php [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+
+    /**
+     * encrypt the Central API Password
+     *
+     * @param string $password
+     * @return string
+     */
+    private static function encryptApiCredentials(string $password): string
+    {
+        if (!isset($_ENV['APP_SECRET'])) {
```
Comment (jeremyjaouen) [inline src/Centreon/Domain/PlatformInformation/Model/PlatformInformationFactory.php:99]: pass throught constructor

File: src/Centreon/Domain/PlatformInformation/UseCase/V20/UpdatePartiallyPlatformInformation.php [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+            $validator->validateOrFail($request);
+        }
+
+        /**
+         * Create Information from Factory to be able to access them independently
+         * and validate the length of each value.
+         * Then transform them as Dto to easily create the PlatformInformation.
+         */
+        $information = InformationFactory::createFromRequest($request);
+        $informationDto = InformationV20Factory::create($information);
```
Comment (jeremyjaouen) [inline src/Centreon/Domain/PlatformInformation/UseCase/V20/UpdatePartiallyPlatformInformation.php:125]: to delete

File: src/Centreon/Domain/PlatformInformation/Model/InformationV20.php [hunk 5]
```diff
@@ -0,0 +1,41 @@
+<?php
+
```
Comment (sc979) [inline src/Centreon/Domain/PlatformInformation/Model/InformationV20.php:2]: to delete

File: src/Centreon/Domain/PlatformInformation/Model/InformationV20Factory.php [hunk 6]
```diff
@@ -0,0 +1,47 @@
+<?php
+
```
Comment (sc979) [inline src/Centreon/Domain/PlatformInformation/Model/InformationV20Factory.php:2]: to delete

File: src/Centreon/Domain/PlatformInformation/UseCase/V20/UpdatePartiallyPlatformInformation.php [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+        $platformInformationUpdate = PlatformInformationFactory::create($informationDto);
+
+        foreach ($information as $informationObject) {
```
Comment (jeremyjaouen) [inline src/Centreon/Domain/PlatformInformation/UseCase/V20/UpdatePartiallyPlatformInformation.php:128]:
```suggestion
        foreach ($informationList as $information) {
```

File: src/Centreon/Domain/PlatformInformation/Model/PlatformInformation.php [hunk 8]
```diff
@@ -328,13 +191,16 @@ public function setApiScheme(?string $schema): self
      * @param int|null $port
      * @return int
```
Comment (sc979) [inline src/Centreon/Domain/PlatformInformation/Model/PlatformInformation.php:192]:
```suggestion
     * @return int|null
```

File: src/Centreon/Domain/PlatformInformation/Model/PlatformInformation.php [hunk 9]
```diff
@@ -370,11 +236,13 @@ public function getApiPath(): ?string
      */
     public function setApiPath(?string $path): self
     {
-        $path = trim(filter_var($path, FILTER_SANITIZE_STRING, ['options' => ['default' => '']]), '/');
-        if (empty($path)) {
-            throw new \InvalidArgumentException(
-                _("Central platform's data are not consistent. Please check the 'Remote Access' form")
-            );
+        if ($path !== null) {
+            $path = trim(filter_var($path, FILTER_SANITIZE_STRING, ['options' => ['default' => '']]), '/');
```
Comment (sc979) [inline src/Centreon/Domain/PlatformInformation/Model/PlatformInformation.php:240]:
```suggestion
            $path = trim(filter_var($path, FILTER_SANITIZE_STRING), '/');
```

File: src/Centreon/Domain/PlatformInformation/UseCase/V20/UpdatePartiallyPlatformInformation.php [hunk 10]
```diff
@@ -0,0 +1,279 @@
...
+
+    /**
+     * Update the Platform Type
+     *
+     * @param PlatformInformation $platformInformationUpdate
+     * @param PlatformInformation $currentPlatformInformation
+     * @throws RemoteServerException
+     */
+    private function updatePlatformTypeOrFail(
+        PlatformInformation $platformInformationUpdate,
```
Comment (sc979) [inline src/Centreon/Domain/PlatformInformation/UseCase/V20/UpdatePartiallyPlatformInformation.php:194]:
```suggestion
        PlatformInformation $platformInformationToUpdate,
```

File: src/Centreon/Domain/RemoteServer/RemoteServerService.php [hunk 11]
```diff
@@ -146,21 +148,27 @@ public function convertCentralToRemote(PlatformInformation $platformInformation)
...
-        }
-
-        try {
             $this->menuRepository->disableCentralMenus();
+        } catch (RepositoryException | PlatformConflictException $ex) {
+            $this->updatePlatformTypeParameters(Platform::TYPE_CENTRAL);
+            throw $ex;
         } catch (\Exception $ex) {
+            $this->updatePlatformTypeParameters(Platform::TYPE_CENTRAL);
             throw new MenuException(_('An error occured while disabling the central menus'));
```
Comment (sc979) [inline src/Centreon/Domain/RemoteServer/RemoteServerService.php:172]:
```suggestion
            throw new MenuException(_('An error occurred while disabling the central menus'));
```

File: src/Centreon/Domain/PlatformInformation/UseCase/V20/UpdatePartiallyPlatformInformation.php [hunk 12]
```diff
@@ -121,27 +120,27 @@ public function execute(array $request): void
...
-            $this->validateCentralServerAddressOrFail($platformInformationUpdate->getCentralServerAddress());
+        if ($platformInformationToUpdate->getCentralServerAddress() !== null) {
+            $this->validateCentralServerAddressOrFail($platformInformationToUpdate->getCentralServerAddress());
         }
 
         $currentPlatformInformation = $this->readRepository->findPlatformInformation();
-
-        if ($platformInformationUpdate->isRemote() !== null) {
-            $this->updatePlatformTypeOrFail($platformInformationUpdate, $currentPlatformInformation);
+        dd($currentPlatformInformation);
```
Comment (sc979) [inline src/Centreon/Domain/PlatformInformation/UseCase/V20/UpdatePartiallyPlatformInformation.php:138]: to delete

File: src/Centreon/Domain/PlatformInformation/UseCase/V20/UpdatePartiallyPlatformInformation.php [hunk 13]
```diff
@@ -121,27 +120,27 @@ public function execute(array $request): void
...
+        if ($platformInformationToUpdate->isRemote() !== null) {
```
Comment (sc979) [inline src/Centreon/Domain/PlatformInformation/UseCase/V20/UpdatePartiallyPlatformInformation.php:139]: shouldn't you check that it's true or false ?

File: src/Centreon/Domain/PlatformInformation/UseCase/V20/UpdatePartiallyPlatformInformation.php [hunk 14]
```diff
@@ -121,27 +120,27 @@ public function execute(array $request): void
          * and validate the length of each value.
          * Then transform them as Dto to easily create the PlatformInformation.
```
Comment (sc979) [inline src/Centreon/Domain/PlatformInformation/UseCase/V20/UpdatePartiallyPlatformInformation.php:121]: Then transform them as Dto to easily create -> Then transform them as Dto to easily update

File: src/Centreon/Domain/PlatformInformation/UseCase/V20/UpdatePartiallyPlatformInformation.php [hunk 15]
```diff
@@ -186,17 +185,17 @@ private function updateProxyOptions(Information $proxyInformation, ?string $cent
...
      * Update the Platform Type
      *
-     * @param PlatformInformation $platformInformationUpdate
+     * @param PlatformInformation $platformInformationToUpdate
      * @param PlatformInformation $currentPlatformInformation
      * @throws RemoteServerException
      */
-    private function updatePlatformTypeOrFail(
-        PlatformInformation $platformInformationUpdate,
+    private function updatePlatformType(
```
Comment (sc979) [inline src/Centreon/Domain/PlatformInformation/UseCase/V20/UpdatePartiallyPlatformInformation.php:192]: the method is too much specific to call it updatePlatformType. Maybe convertCentralToRemote suits it best

File: src/Centreon/Domain/PlatformInformation/UseCase/V20/UpdatePartiallyPlatformInformation.php (hunk 14 above)
Comment (jeremyjaouen) [reply src/Centreon/Domain/PlatformInformation/UseCase/V20/UpdatePartiallyPlatformInformation.php:121]: as we no more use this transition Dto, i'll deleted the line.

File: src/Centreon/Domain/PlatformInformation/UseCase/V20/UpdatePartiallyPlatformInformation.php (hunk 15 above)
Comment (jeremyjaouen) [reply src/Centreon/Domain/PlatformInformation/UseCase/V20/UpdatePartiallyPlatformInformation.php:191]: depending of the PlatformInformation to update and the current PlatformInformation, this method could convert Central to Remote or a Remote to Central.

Comment (sc979) [reply src/Centreon/Domain/PlatformInformation/UseCase/V20/UpdatePartiallyPlatformInformation.php:191]: the intention is not reflected in the method name... As you will not be able to change the type or platforms which are not remote or central (eg : poller and coming MBI / MAP / deportedDB / centreon-cloud)

Comment (jeremyjaouen) [reply src/Centreon/Domain/PlatformInformation/UseCase/V20/UpdatePartiallyPlatformInformation.php:191]: agreed, we should discuss further on how to name this method with a relevant name that is not too long.

**Review by sc979 — APPROVED** [review]

**Review by loiclau — APPROVED** [review]

File: src/Centreon/Domain/PlatformInformation/Model/PlatformInformation.php [hunk 16]
```diff
@@ -144,29 +86,9 @@ public function isRemote(): bool
      * @param bool $isRemote
      * @return $this
      */
-    public function setIsRemote(bool $isRemote): self
+    public function setRemote(?bool $isRemote): self
```
[the remark this answers is not in the document]
Comment (jeremyjaouen) [reply src/Centreon/Domain/PlatformInformation/Model/PlatformInformation.php:89]: Not really, as if I patch other information (maybe just the ApiUsername), all the other properties has to be null to be sure to not override them.

File: src/Centreon/Infrastructure/PlatformInformation/Repository/Model/PlatformInformationFactoryRDB.php [hunk 17]
```diff
@@ -0,0 +1,114 @@
...
+     */
+    private function decryptApiCredentials(?string $encryptedKey): ?string
+    {
+        if (empty($encryptedKey)) {
+            return null;
+        }
+
+        if ($this->appSecret === null) {
+            throw new \InvalidArgumentException(
+                _("Unable to find the encryption key. Please check the '.env.local.php' file.")
```
Comment (callapa) [inline src/Centreon/Infrastructure/PlatformInformation/Repository/Model/PlatformInformationFactoryRDB.php:97]:
```suggestion
                _("Unable to find the encryption key")
```

File: src/Centreon/Infrastructure/PlatformInformation/Repository/Model/PlatformInformationFactoryRDB.php [hunk 18]
```diff
@@ -0,0 +1,114 @@
...
+            );
+        }
+
+        // second key
+        $secondKey = base64_encode(self::ENCRYPT_SECOND_KEY);
+
+        try {
+            $centreonEncryption = new Encryption();
+            $centreonEncryption->setFirstKey($_ENV['APP_SECRET'])->setSecondKey($secondKey);
```
Comment (callapa) [inline src/Centreon/Infrastructure/PlatformInformation/Repository/Model/PlatformInformationFactoryRDB.php:106]: Do not use environnement variables

File: src/Centreon/Infrastructure/PlatformInformation/Repository/PlatformInformationRepositoryRDB.php [hunk 19]
```diff
@@ -0,0 +1,226 @@
...
+                    $row['value'] = $row['value'] === 'yes';
+                }
+
+                $result[$row['key']] = $row['value'];
+            }
+            if (!empty($result)) {
+                /**
+                 * @var PlatformInformation $platformInformation
+                 */
+                $platformInformationFactoryRDB = new PlatformInformationFactoryRDB($_ENV['APP_SECRET']);
```
Comment (callapa) [inline src/Centreon/Infrastructure/PlatformInformation/Repository/PlatformInformationRepositoryRDB.php:81]: Do not use environnement variables

**Review by callapa — CHANGES_REQUESTED** [review]

File: src/Centreon/Domain/PlatformInformation/UseCase/V20/UpdatePartiallyPlatformInformation.php [hunk 20]
```diff
@@ -0,0 +1,323 @@
...
+                sprintf(
+                    _('the address %s is already used has proxy address and can\'t ' .
+                    'be provided as Central Server Address'),
+                    $centralServerAddress
+                )
+            );
+        }
+    }
+
+    private function populateMissingInformationValues(
```
Comment (kduret) [inline src/Centreon/Domain/PlatformInformation/UseCase/V20/UpdatePartiallyPlatformInformation.php:283]: could you add phpdoc

**Review by kduret — APPROVED** [review]

File: src/Centreon/Domain/PlatformInformation/Model/Information.php [hunk 21]
```diff
@@ -0,0 +1,96 @@
...
+    }
+
+    /**
+     * @param string $key
+     * @throws \Assert\AssertionFailedException
+     * @return self
+     */
+    public function setKey(string $key): self
+    {
+        Assertion::minLength($key, self::MIN_LENGTH, 'Information::key');
```
Comment (callapa) [inline src/Centreon/Domain/PlatformInformation/Model/Information.php:68]:
```suggestion
        Assertion::minLength($key, self::MIN_KEY_LENGTH, 'Information::key');
```

File: src/Centreon/Domain/PlatformInformation/Model/Information.php [hunk 22]
```diff
@@ -0,0 +1,96 @@
...
+
+    /**
+     * @param mixed $value
+     * @throws \Assert\AssertionFailedException
+     * @return self
+     */
+    public function setValue($value): self
+    {
+        if ($value !== null && is_string($value)) {
+            Assertion::minLength($value, self::MIN_LENGTH, 'Information::value');
```
Comment (callapa) [inline src/Centreon/Domain/PlatformInformation/Model/Information.php:90]: Are you sure that it is not possible to set the value to null or empty?

File: tests/php/Centreon/Domain/PlatformInformation/Model/InformationTest.php [hunk 23]
```diff
@@ -0,0 +1,95 @@
...
+use PHPUnit\Framework\TestCase;
+use Centreon\Domain\Common\Assertion\AssertionException;
+use Centreon\Domain\PlatformInformation\Model\Information;
+
+class InformationTest extends TestCase
+{
+    /**
+     * Too long Key Test.
+     */
+    public function testKeyTooLongException(): void
```
Comment (callapa) [inline tests/php/Centreon/Domain/PlatformInformation/Model/InformationTest.php:35]: Don't forget to add the test testKeyTooShortException

**Review by callapa — CHANGES_REQUESTED** [review]

**Review by callapa — APPROVED** [review]

**Pull request merged**